pkg/kubelet/container/interface.go

```diff
@@ -0,0 +1,296 @@
+// Sandboxes.
+type SandboxManager interface {
+	// Create creates a sandbox based on the given config, and returns the ID
+	// of the new sandbox.
+	Create(config *SandboxConfig) (string, error)
+	// Delete deletes the sandbox with by its ID. If there are any running
+	// containers in the sandbox, they will be terminated as a side-effect.
+	Delete(id string) (string, error)
+	// List lists existing sandboxes, filtered by the given SandboxFilter.
+	List(filter SandboxFilter) []SandboxListItem
+	// Inspect gets the detailed config and status of the sandbox by ID.
+	Inspect(id string) Sandbox
+}
+
+// SandboxConfig holds all the required and optional fields for creating a
+// sandbox.
+type SandboxConfig struct {
+	// Name is the name of the sandbox.
+	Name string
+	// Hostname is the hostname of the sandbox.
+	Hostname string
+	// DNSOptions the DNS configuration the sandbox.
+	DNSOptions *DNSOptions
+	// PortMappings lists the port mappings for the sandbox.
+	PortMappings []PortMapping
+	// Mounts list the mounts to be added to the sandbox's filesystem.
+	Mounts []Mount
+	// Resources specifies the resource requirements for the sandbox.
+	Resources *Resources
+	// Annotations is an unstructured key value map that may be set by external
+	// tools to store and retrieve arbitrary metadata.
+	Annotations map[string]string
+	// Linux contains configurations specific to Linux containers.
+	Linux *LinuxSandboxConfig
+}
+
+// LinuxSandboxConfig holds platform-specific configuraions for Linux
+// containers.
+type LinuxSandboxConfig struct {
+	// CgroupParent is the parent cgroup of the sandbox.
+	CgroupParent string
+	// NamespaceOptions contains configurations for the sanbox's namespaces.
+	NamespaceOptions NamespaceOptions
+	// SELinux is the SELinux context to be applied to all containers.
+	SELinux *api.SELinuxOptions
+}
+
+// NamespaceOptions provides options for Linux namespaces.
+type NamespaceOptions struct {
+	// HostNetwork uses the host's network namespace.
+	HostNetwork bool
+	// HostPID uses the host's pid namesapce.
+	HostPID bool
+	// HostIPC uses the host's ipc namespace.
+	HostIPC bool
+}
+
+// DNSOptions specifies the DNS servers and search domains.
+type DNSOptions struct {
+	// Servers is a list of DNS servers of the cluster.
+	Servers []string
+	// Searches is a list of DNS search domains of the cluster.
+	Searchs []string
+}
+
+type SandboxState string
+
+const (
+	// SandboxActive means the sandbox is functioning properly.
+	SandboxActive SandboxState = "active"
+	// SandboxInactive means the sandbox is not functioning properly.
+	SandboxInactive SandboxState = "inactive"
+)
+
+// SandboxFilter is used to filter a list of Sandboxes.
+type SandboxFilter struct {
+	// Name of the sandbox.
+	Name string
+	// ID of the sandbox.
+	ID string
+	// State of the sandbox.
+	State string
+	// Annotations of the sandbox. A container needs to have all listed
+	// annotations to be sandbox a match.
+	Annotations map[string]string
```

Comment:
If we want to stick to Kube API semantics, we should split up Annotations
from labels. If grouping is our goal, then we should specify Labels and use
that.

On Mon, May 16, 2016 at 4:12 PM, Yu-Ju Hong <notifications@github.com>
wrote:

> In pkg/kubelet/container/interface.go
> <https://github.com/yujuhong/kubernetes/pull/1#discussion_r63441769>:
>
> > +	SandboxActive SandboxState = "active"
> > +	// SandboxInactive means the sandbox is not functioning properly.
> > +	SandboxInactive SandboxState = "inactive"
> > +)
> > +
> > +// SandboxFilter is used to filter a list of Sandboxes.
> > +type SandboxFilter struct {
> > +	// Name of the sandbox.
> > +	Name string
> > +	// ID of the sandbox.
> > +	ID string
> > +	// State of the sandbox.
> > +	State string
> > +	// Annotations of the sandbox. A container needs to have all listed
> > +	// annotations to be sandbox a match.
> > +	Annotations map[string]string
>
> We implement labels through annotations. This is required for grouping
> containers by pod UID, for example.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/yujuhong/kubernetes/pull/1/files/ddd041c68ca94d95fa591dce703932fe5d577dd0#r63441769>
>